File: server/session.ts
import { randomBytes } from 'crypto';

export interface Session {
  sessionId: string;
  createdAt: Date;
  lastAccessedAt: Date;
  ip?: string;
}

class SessionStore {
  private sessions: Map<string, Session> = new Map();
  private readonly sessionDuration = 30 * 60 * 1000; // 30 minutes

  generateSessionId(): string {
    return randomBytes(32).toString('hex');
  }

  createSession(ip?: string): Session {
    const sessionId = this.generateSessionId();
    const session: Session = {
      sessionId,
      createdAt: new Date(),
      lastAccessedAt: new Date(),
      ip,
    };
    this.sessions.set(sessionId, session);
    this.cleanupExpiredSessions();
    return session;
  }

  getSession(sessionId: string): Session | null {
    const session = this.sessions.get(sessionId);
    if (!session) return null;

    if (this.isSessionExpired(session)) {
      this.sessions.delete(sessionId);
      return null;
    }

    session.lastAccessedAt = new Date();
    return session;
  }

  deleteSession(sessionId: string): boolean {
    return this.sessions.delete(sessionId);
  }

  private isSessionExpired(session: Session): boolean {
    const now = new Date().getTime();
    const lastAccessed = session.lastAccessedAt.getTime();
    return now - lastAccessed > this.sessionDuration;
  }

  private cleanupExpiredSessions(): void {
    const now = new Date().getTime();
    const entries = Array.from(this.sessions.entries());
    for (const [sessionId, session] of entries) {
      const lastAccessed = session.lastAccessedAt.getTime();
      if (now - lastAccessed > this.sessionDuration) {
        this.sessions.delete(sessionId);
      }
    }
  }

  getAllSessions(): Session[] {
    return Array.from(this.sessions.values());
  }
}

export const sessionStore = new SessionStore();

File: go-server/internal/middleware/session.go
package middleware

import (
	"crypto/rand"
	"encoding/hex"
	"sync"
	"time"
)

type Session struct {
	SessionID      string    `json:"sessionId"`
	CreatedAt      time.Time `json:"createdAt"`
	LastAccessedAt time.Time `json:"lastAccessedAt"`
	IP             string    `json:"ip,omitempty"`
}

// Sessions defines the interface for session management.
// Both SessionStore (in-memory) and RedisSessionStore implement this.
type Sessions interface {
	CreateSession(ip string) *Session
	GetSession(sessionID string) *Session
	DeleteSession(sessionID string) bool
}

// generateSessionID produces a cryptographically random 64-character hex string.
func generateSessionID() string {
	b := make([]byte, 32)
	if _, err := rand.Read(b); err != nil {
		panic("crypto/rand failed: " + err.Error())
	}
	return hex.EncodeToString(b)
}

type SessionStore struct {
	mu              sync.Mutex
	sessions        map[string]*Session
	sessionDuration time.Duration
}

func NewSessionStore(maxAge time.Duration) *SessionStore {
	return &SessionStore{
		sessions:        make(map[string]*Session),
		sessionDuration: maxAge,
	}
}

func (s *SessionStore) CreateSession(ip string) *Session {
	s.mu.Lock()
	defer s.mu.Unlock()

	sessionID := generateSessionID()
	now := time.Now()
	session := &Session{
		SessionID:      sessionID,
		CreatedAt:      now,
		LastAccessedAt: now,
		IP:             ip,
	}
	s.sessions[sessionID] = session
	s.cleanupExpired()
	return session
}

func (s *SessionStore) GetSession(sessionID string) *Session {
	s.mu.Lock()
	defer s.mu.Unlock()

	session, ok := s.sessions[sessionID]
	if !ok {
		return nil
	}

	if s.isExpired(session) {
		delete(s.sessions, sessionID)
		return nil
	}

	session.LastAccessedAt = time.Now()
	return session
}

func (s *SessionStore) DeleteSession(sessionID string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()

	_, ok := s.sessions[sessionID]
	if ok {
		delete(s.sessions, sessionID)
	}
	return ok
}

func (s *SessionStore) isExpired(session *Session) bool {
	return time.Since(session.LastAccessedAt) > s.sessionDuration
}

func (s *SessionStore) cleanupExpired() {
	now := time.Now()
	for id, session := range s.sessions {
		if now.Sub(session.LastAccessedAt) > s.sessionDuration {
			delete(s.sessions, id)
		}
	}
}
